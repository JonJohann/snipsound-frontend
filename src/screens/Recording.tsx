import React, { useEffect, useState } from 'react';
import { View, TouchableOpacity, Text } from 'react-native';
import { Audio } from 'expo-av';
const axios = require('axios').default;
import * as FileSystem from 'expo-file-system';
import { AppState } from '../screens/SoundList'
import { useSelector, useDispatch } from 'react-redux';
import TitleDescription from '../components/recording/TitleDescription'
import CategoryRow from '../components/recording/CategoryRow'
import SnipsoundLogo from '../components/misc/SnipsoundLogo'
import { resetParams } from '../state/ParamsSlice';
import RecordingRow from '../components/recording/RecordingRow';
import Divider from '../components/misc/Divider';

var initialSound = new Audio.Sound()
var initialRecord = new Audio.Recording();

// Screen for the recording of sound files
export default function Recording({ navigation }: any) {

    const [title, setTitle] = useState("")
    const [category, setCategory] = useState("Rap")
    const [description, setDescription] = useState("")
    const owner = useSelector((state: AppState) => state.sounds.user)

    const dispatch = useDispatch()

    const [sound, setSound] = useState(initialSound);
    const [isPlaying, setIsPlaying] = useState(false);

    const [isRecording, setIsRecording] = useState(false);
    const [recording, setRecording] = useState(initialRecord);
    const [songReady, setSongReady] = useState(false);

    useEffect(() => {
        Audio.requestPermissionsAsync()
        var initialSound = new Audio.Sound()
        var initialRecord = new Audio.Recording();
        setSound(initialSound);
        setRecording(initialRecord)
    },[])

    const canPost = () => {
        const titleBool = title.length >= 1 && title.length < 13
        const descriptionBool = description.length < 180
        return titleBool && descriptionBool
    }

    const recordSound = async () => {

        const status = await recording.getStatusAsync();
        try {
            if (!status.isRecording) {
                await Audio.setAudioModeAsync({
                    allowsRecordingIOS: true,
                    interruptionModeIOS: Audio.INTERRUPTION_MODE_IOS_DO_NOT_MIX,
                    playsInSilentModeIOS: true,
                    shouldDuckAndroid: true,
                    interruptionModeAndroid: Audio.INTERRUPTION_MODE_ANDROID_DO_NOT_MIX,
                    playThroughEarpieceAndroid: false,
                    staysActiveInBackground: true,
                });
                await recording.prepareToRecordAsync(Audio.RECORDING_OPTIONS_PRESET_HIGH_QUALITY)
                await recording.startAsync();
                setIsRecording(true);

            }
        } catch (error) {
            console.log(error)

        }
        if (isRecording) {
            stopRecord()

        }
    }

    const stopRecord = async () => {
        try {
            await Audio.setAudioModeAsync({
                allowsRecordingIOS: false,
                interruptionModeIOS: Audio.INTERRUPTION_MODE_IOS_DO_NOT_MIX,
                playsInSilentModeIOS: true,
                shouldDuckAndroid: true,
                interruptionModeAndroid: Audio.INTERRUPTION_MODE_ANDROID_DO_NOT_MIX,
                playThroughEarpieceAndroid: false,
                staysActiveInBackground: true,
            });
            await recording.stopAndUnloadAsync();
            await recording.createNewLoadedSoundAsync().then(sounds => {
                setSound(sounds.sound)
            })
            setIsRecording(false);
            setSongReady(true)
        } catch (error) {
            console.log(error)
        }
    }

    const newRecord = async () => {
        if (isPlaying) {
            playSound()
        }
        initialRecord = new Audio.Recording();
        setRecording(initialRecord)
        setSongReady(false);
    }

    const playSound = async () => {

        if (!isPlaying) {

            await sound.setIsLoopingAsync(true)
            await sound.playAsync();
            setIsPlaying(true);
        }
        else {

            await sound.stopAsync();
            setIsPlaying(false);
        }
    }

    const saveSound = async () => {
        const uri = String(recording.getURI());
        const url = "http://84.202.202.48:3074/upload";
        const snipsound = {
            title: title,
            date: Date.now(),
            owner: owner,
            category: category,
            filename: "",
            description: description,
            likes: 0
        }
        if (isPlaying) {
            playSound()
        }

        initialSound = new Audio.Sound()
        initialRecord = new Audio.Recording();

        FileSystem.uploadAsync(url, uri).then(res => {
            snipsound.filename = res.body
            axios.post("http://84.202.202.48:3074/", snipsound)
        }).then(() =>  {dispatch(resetParams()); navigation.navigate("SoundList")})
    }

    const recordingRowProps = {
        recordSound,
        newRecord,
        playSound,
        saveSound,
        canPost,
        songReady,
        isPlaying, 
        isRecording
    }

    return (
        <View style={{ flex: 1, alignItems: "center", justifyContent: "center" }}>
            <SnipsoundLogo />
            <View style={{ flex: 2 }}>
                <CategoryRow categoryChosen={setCategory} />
                {title.length > 12 ? <Text style={{color: "red"}}>Title must be less than 13 characters...</Text>: null}
                <TitleDescription setTitle={setTitle} setDescription={setDescription} />
                {description.length > 179 ? <Text style={{color: "red"}}>Description must be less than 180 characters...</Text>: null}
                <RecordingRow {...recordingRowProps}/>
            </View>
        </View>

    );
}