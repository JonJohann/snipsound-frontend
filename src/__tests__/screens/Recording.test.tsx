import React from 'react';
import renderer, { ReactTestRendererJSON } from 'react-test-renderer';
import Recording from '../../screens/Recording';
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'

describe('<Recording />', () => {
    const initialState = {
        sounds: {
            user: ""
        }
    }
    beforeEach(() => {
    })
    const mockStore = configureStore()
    let store = mockStore(initialState)
    it('has 2 children', () => {
        const tree = renderer.create(<Provider store={store}><Recording /></Provider>).toJSON() as ReactTestRendererJSON;
        expect(tree!.children!.length).toBe(2);
    });

    it('renders correctly', () => {
        const tree = renderer.create(<Provider store={store}><Recording /></Provider>).toJSON();
        expect(tree).toMatchSnapshot();
    });
});