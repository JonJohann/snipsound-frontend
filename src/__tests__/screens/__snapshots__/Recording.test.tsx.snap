// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Recording /> renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Image
    resizeMode="contain"
    source={1}
    style={
      Object {
        "alignSelf": "stretch",
        "marginHorizontal": 10,
        "marginVertical": -110,
        "width": undefined,
      }
    }
  />
  <View
    style={
      Object {
        "flex": 2,
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          Object {
            "alignSelf": "center",
            "fontSize": 20,
            "marginRight": 5,
          }
        }
      >
        Category:
      </Text>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "white",
            "borderColor": "#F194FF",
            "borderRadius": 10,
            "borderWidth": 1,
            "elevation": 2,
            "marginHorizontal": 2,
            "marginVertical": 5,
            "opacity": 1,
            "paddingHorizontal": 12,
            "paddingVertical": 8,
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.25,
            "shadowRadius": 0.84,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "black",
              "fontFamily": "Manrope_400Regular",
              "fontSize": 20,
            }
          }
        >
          Rap
        </Text>
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#e6e6e6",
            "borderColor": "#6d557000",
            "borderRadius": 10,
            "borderWidth": 1,
            "elevation": 2,
            "marginHorizontal": 2,
            "marginVertical": 5,
            "opacity": 1,
            "paddingHorizontal": 12,
            "paddingVertical": 8,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "grey",
              "fontFamily": "Manrope_400Regular",
              "fontSize": 20,
            }
          }
        >
          Song
        </Text>
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#e6e6e6",
            "borderColor": "#6d557000",
            "borderRadius": 10,
            "borderWidth": 1,
            "elevation": 2,
            "marginHorizontal": 2,
            "marginVertical": 5,
            "opacity": 1,
            "paddingHorizontal": 12,
            "paddingVertical": 8,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "grey",
              "fontFamily": "Manrope_400Regular",
              "fontSize": 20,
            }
          }
        >
          Spoken
        </Text>
      </View>
    </View>
    <View>
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="Title"
          rejectResponderTermination={true}
          style={
            Array [
              Object {
                "alignItems": "center",
                "borderBottomWidth": 0.5,
                "borderColor": "gray",
                "borderRadius": 10,
                "fontSize": 20,
                "height": 40,
                "marginLeft": 10,
                "marginTop": 10,
                "paddingHorizontal": 10,
              },
              Object {
                "flex": 1,
              },
            ]
          }
          underlineColorAndroid="transparent"
        />
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "margin": 10,
              "marginLeft": 2,
              "marginRight": -1,
              "marginTop": 15,
              "opacity": 1,
            }
          }
        >
          <Text />
        </View>
      </View>
      <View
        style={
          Object {
            "width": 350,
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          multiline={true}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="Description"
          rejectResponderTermination={true}
          style={
            Object {
              "alignItems": "center",
              "borderBottomWidth": 0.5,
              "borderColor": "gray",
              "borderRadius": 10,
              "fontFamily": "Manrope_400Regular",
              "fontSize": 20,
              "height": 100,
              "marginLeft": 10,
              "marginTop": 10,
              "paddingHorizontal": 10,
            }
          }
          underlineColorAndroid="transparent"
        />
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "center",
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Text />
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Text />
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "margin": 10,
            "opacity": 1,
          }
        }
      >
        <Text />
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignSelf": "center",
            "backgroundColor": "#e6e6e6",
            "borderColor": "#6d557000",
            "borderRadius": 10,
            "borderWidth": 1,
            "elevation": 2,
            "marginHorizontal": 5,
            "marginTop": 10,
            "marginVertical": 5,
            "opacity": 1,
            "paddingHorizontal": 10,
            "paddingVertical": 8,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 20,
            }
          }
        >
          Upload to Snipsound
        </Text>
      </View>
    </View>
  </View>
</View>
`;
